Add persistent folding to Animation Library Editor

Illauriel opened the pull request:
- Resolves the second part of godotengine/godot-proposals#8226
- Supersedes #84088

Enables folding of the library entries in Animation Library Editor, necessary on big projects with hundreds of animations to prevent excessive scrolling.

For folding to be persistent, the `collapsed` state of TreeItems is saved in a subfolder of .godot/editor/ or wherever get_project_settings_dir() would point. I'd love to do it via editor_folding.ccp, but I couldn't find out how to repurpose existing methods from that class, and didn't want to disrupt anything by accident. Because of that all file manipulations are performed in animation_library_editor.cpp.

Due to the fact that multiple AnimationMixers can be present in a scene and share same libraries, the persistence files are created per mixer node path and not per scene. Filename is formed as (pseudocode):
```"folding-" + (scene_file_path + mixer_node_path).md5()+".cfg"```
Because of this, clutter of outdated files can form when the scene filename or the mixers' nodepaths change.

I would be glad if someone more knowledgeable could suggest how to do persistence in editor more gracefully.^^'

Illauriel committed:
Add persistent folding to Animation Library Editor

fire reviewed (approved):
@TokageItLab I think adding persistent folding to the animation library editor is fine, but I also want your review.

Illauriel commented on editor/plugins/animation_library_editor.cpp:
```diff
@@ -29,14 +29,18 @@
 /**************************************************************************/
 
 #include "animation_library_editor.h"
+#include "core/io/file_access.h"
```
Wow, thanks! I was sure I need it to use FileAccess, but it compiles without it x) And no it's not my editor, it's all my fault.^^' 

I'll remove the include in a bit.

KoBeWi commented:
`lib_fold` should be a single file, not a whole folder. You can make it a ConfigFile and use sections for each AnimationMixer, instead of files.

Illauriel commented:
> `lib_fold` should be a single file, not a whole folder. You can make it a ConfigFile and use sections for each AnimationMixer, instead of files.

Sounds reasonable. I was kinda mimicking how scene view folding has separate file for each scene, but it irked me that it creates clutter. I'll do an attempt to change it tonight :)

Illauriel commented:
Changed `_load_lib_folding()` and `_update_lib_folding()` to use a single `lib_folding.cfg()` as per KoBeWi's suggestion. Removed `_create_fold_dir()`.
Hope it looks a little cleaner now ^^'

TokageItLab reviewed (dismissed):
I agree with the implementation of folding, and using loosely coupling to implement it with the actual object in the scene is good from a maintainability standpoint.

On the other hand, there is a problem of unstable behavior when renaming, since the path for storing the folding state depends on the name.

Renaming the AnimationMixer in the SceneTree clears all the folding state. In addition, if the names of the two AnimationMixers are switched, the state of the AnimationLibrary with same name will be strange.

![image](https://github.com/godotengine/godot/assets/61938263/3f27c93e-93f0-4547-a611-baf13b5a3c64)

When renaming an AnimationLibrary in the folding state, it expands folding at the moment of renaming and collapses folding when renamed again to the original name.

![image](https://github.com/godotengine/godot/assets/61938263/438cf3a2-206c-4de7-adcc-0e861e3b3cb1)

I think it would be fine to store the state by path by name when saving, but it would be necessary to tie them together with ObjectID or pointer while the editor is running. In other words, it should have paths by ObjectID while the editor is running and needs to add methods that will convert the path between name and ObjectID when the scene is opened and when it is saved.

Illauriel commented:
> On the other hand, there is a problem of unstable behavior when renaming, since the path for storing the folding state depends on the name.

You're right! I never figured out how to do it correctly so that renames and changing Mixer's place in hierarchy doesn't affect it, but using ObjectID sounds promising! I'll try^^

Illauriel commented:
> The ObjectID is updated by memory allocation and should not be stored and therefore needs to be converted with the name.

So over the night, I've managed to solve problem #2 where renaming libraries was causing fold state to revert to unfolded.
I still need to solve #1 where renaming the mixer itself causes loss of saved state, and to decide how to cleanup old names from the save file. Should be doable. ^^'

fire reviewed (approved):
Looks good.

I contributed the signature idea to @Illauriel, and we discussed the idea at the animation meeting.

I still need to test the code.

Illauriel commented:
> Looks good.
> 
> I contributed the signature idea to @Illauriel, and we discussed the idea at the animation meeting.
> 
> I still need to test the code.

Thanks! 

I actually found another corner case, but it's a quick thing to fix. Gonna make an edit in a few min.

Illauriel commented:
Okay, I think now we're good. Phew, it's been a while!

> On the other hand, there is a problem of unstable behavior when renaming, since the path for storing the folding state depends on the name.

I addressed all cases where editing names of Mixers, names of Libraries, or moving Mixer in the hierarchy would lose folding state.

> I think it would be fine to store the state by path by name when saving, but it would be necessary to tie them together with ObjectID or pointer while the editor is running.

I followed this suggestion, and started storing both names and ObjectIDs for current session. If ObjectIDs are outdated (Usually because it's a new session) it falls back on names again. In case the hierarchy has changed, and some other cases, it tries to recover data, looking for current mixer's ObjectID in other saved entries. Another fallback that @fire mentioned above is so called "mixer signature" - a hash of all the libraries and animations in the mixer. It should be sufficient to recover for all remaining cases, unless the library list is edited via code. But since in that case user doesn't interact with AnimationLibraryEditor at all, folding shouldn't matter all that much. 
Also the system tries its best to cleanup and actualize outdated entries.

SaracenOne reviewed (approved):
Was discussed and demonstrated in the Animation meeting and seems fine 👍

Repiteo commented on editor/plugins/animation_library_editor.h:
```diff
@@ -31,11 +31,14 @@
 #include "scene/animation/animation_mixer.h"
 #include "scene/gui/dialogs.h"
 #include "scene/gui/tree.h"
+#include <cstdint>
```
```suggestion
```

Illauriel commented on editor/plugins/animation_library_editor.cpp:
```diff
@@ -30,13 +30,20 @@
 #include "editor/themes/editor_scale.h"
 #include "scene/animation/animation_mixer.h"
+#include <cstddef>
+#include <cstdint>
```
Made changes, rebuilt, tested, seems like nothing exploded x) 

Thanks for the correction^^

Repiteo commented:
Thanks!
